src/model_clay.py

```diff
@@ -892,13 +895,28 @@ def predict_step(
         )
         assert not torch.isnan(embeddings_raw).any()  # ensure no NaNs in embedding
 
-        # Take the mean of the embeddings along the sequence_length dimension
-        # excluding the last two latlon_ and time_ embeddings, i.e. compute
-        # mean over patch embeddings only
-        embeddings_mean: torch.Tensor = embeddings_raw[:, :-2, :].mean(dim=1)
-        assert embeddings_mean.shape == torch.Size(
-            [self.model.encoder.B, 768]  # (batch_size, hidden_size)
-        )
+        if self.hparams.output_patch_embeddings:
+            # Take the mean of the embeddings along the group dimension
+            # excluding the last two latlon_ and time_ embeddings. This
+            # results in one embedding per patch.
+            embeddings_raw = rearrange(
+                embeddings_raw[:, :-2, :], "b (g l) s -> b g (l s)", l=256, g=6
+            )
+            embeddings_mean = reduce(embeddings_raw, "b g s -> b s", "mean")
+            assert embeddings_mean.shape == torch.Size(
+                [
+                    self.model.encoder.B,
+                    256 * 768,
+                ]  # (batch_size, nr of patches * hidden_size)
+            )
```

Comment:
Unsqueezed would undoubtedly be better, but I assumed that we can only store one-dimensional arrays in a field. I guess from your questions that is not the case. Can we store multidimensional arrays here? If yes, lets store the unsqueezed one for sure!

Reply:
Yes, GeoParquet/Arrow technically supports [FixedShapeTensorArray](https://arrow.apache.org/docs/14.0/python/generated/pyarrow.FixedShapeTensorArray.html) which can be multi-dimensional.

Reply:
That will require a rewrite without pandas, because pandas can not handle this and implicitly flattens the `FixedShapeTensorArray`, see example below. I ran across this when running tests and not getting the expected 2D arrays. This defintively complicates things in terms of keeping the structure.

```
In [15]: array = np.arange(8).reshape((2,2,2))
    ...: 
    ...: arrow = pa.FixedShapeTensorArray.from_numpy_ndarray(array)
    ...: 
    ...: df = pd.DataFrame(data={"arrow": arrow})
    ...: 
    ...: df
    ...: 
Out[15]: 
          arrow
0  [0, 1, 2, 3]
1  [4, 5, 6, 7]
```

Reply:
Ok so based on the above, the arrays are outputted in a flat structure for now. I updated the documentation with a sentence about where the structure comes from and why the arrays are flat.

https://github.com/Clay-foundation/model/pull/133/files#diff-2f39d0012d2540cdab1ac1ff882a3875d829b85f0acee68fe102e6b8a4f2c6c0R60

This is not ideal but keeping the multidimensional arrays now would require refactoring of how we construct the geoparquet files.

Reply:
Hmm, have you tried inserting a FixedShapeTensorArray into a row cell directly? This seems to work:

```python
import numpy as np
import pandas as pd
import pyarrow as pa

array0 = np.arange(0, 8).reshape((2, 2, 2))
array1 = np.arange(9, 17).reshape((2, 2, 2))

arrow0 = pa.FixedShapeTensorArray.from_numpy_ndarray(array0)
arrow1 = pa.FixedShapeTensorArray.from_numpy_ndarray(array1)

df = pd.DataFrame(data={"embeddings": [arrow0, arrow1]})
print(df)
#                             embeddings
# 0         ([0, 1, 2, 3], [4, 5, 6, 7])
# 1  ([9, 10, 11, 12], [13, 14, 15, 16])
```

Reply:
The arrow part still flattens the arrays that are input. For this to work, we would have to do the manual deconstruction up to 3 dimensions deep. Not sure that this will make things more user friendly. With this approach, the user would have to reconstruct a list of a list of a list of arrays in the "group" case.